### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by oleg1std**

### CODE REVIEW

File: HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+            if (objects.Count > MaxObjectsNumberToUpload)
+                Result.Failure<List<string>>($"Can't upload more than {MaxObjectsNumberToUpload} objects at one time");
+
+            var processedTasks = new List<Result<string>>(objects.Count);
+            var nextTaskIndex = 0;
+            var processingTasks = new List<Task<Result<string>>>(objects.Count < UploadConcurrencyNumber
+                ? objects.Count
+                : UploadConcurrencyNumber);
+            
+            for (; nextTaskIndex < UploadConcurrencyNumber && nextTaskIndex < objects.Count; nextTaskIndex++)
```
Comment (kirillta) [inline HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs:86]: Please use the full form

File: HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+            {
+                var objectToUpload = objects[nextTaskIndex];
+                processingTasks.Add(Add(objectToUpload.key, objectToUpload.stream, cancellationToken));
```
Comment (kirillta) [inline HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs:89]: Add(Add)?

Comment (kirillta) [reply HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs:89]: That's hard to read

File: HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        }
+        
+        
+        private readonly ILogger<AmazonS3ClientService> _logger;
+
+        private const string PathString = "https://s3.{0}.amazonaws.com/{1}/{2}";
+        private readonly Amazon.S3.AmazonS3Client _s3Client;
+        private readonly AmazonS3ClientOptions _options;
+        private const int MaxObjectsNumberToUpload = 50;
+        private const int UploadConcurrencyNumber = 5;
```
Comment (kirillta) [inline HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs:239]: Why 5? Shouldn't that depend on the proc count?

File: HappyTravel.AmazonS3Client/Infrastructure/Logging/LoggerExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using Microsoft.Extensions.Logging;
+
+namespace HappyTravel.AmazonS3Client.Infrastructure.Logging
+{
+    internal static class LoggerExtensions
+    {
+        static LoggerExtensions()
+        {
+            S3RequestExceptionOccured = LoggerMessage.Define(LogLevel.Error,
```
Comment (kirillta) [inline HappyTravel.AmazonS3Client/Infrastructure/Logging/LoggerExtensions.cs:10]: We have a package for exception generation [here](https://github.com/happy-travel/logger-message-generator)

File: HappyTravel.AmazonS3Client/Options/AmazonS3ClientOptions.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    public class AmazonS3ClientOptions
+    {
+        public string AccessKeyId { get; set; } = string.Empty;
+        
+        public string AccessKey { get; set; } = string.Empty;
+        
+        /// <summary>
+        /// Name of an existed bucket
+        /// </summary>
+        public string BucketName { get; set; } = string.Empty;
```
Comment (kirillta) [inline HappyTravel.AmazonS3Client/Options/AmazonS3ClientOptions.cs:15]: I consider remove the bucket name from the options. We will use at least two buckets from the very start: one for images and another for documents

File: HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    {
+        public AmazonS3ClientService(IOptions<AmazonS3ClientOptions> options, ILogger<AmazonS3ClientService> logger)
+        {
+            _options = options.Value;
+            _logger = logger;
+            _s3Client = new Amazon.S3.AmazonS3Client(_options.AccessKeyId, _options.AccessKey, _options.AmazonS3Config);
+        }
+
+
+        /// <summary>
```
Comment (mingazhev) [inline HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs:28]: So bucket name should be added there

File: HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        /// <param name="key">A name of the object in the bucket. Can include subfolders: folder/file1.jpg </param>
+        /// <param name="stream"></param>
+        /// <param name="acl">Access control lists https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html#canned-acl</param>
+        /// <param name="cancellationToken"></param>
+        /// <returns>Url to the object</returns>
+        public async Task<Result<string>> Add(string key, Stream stream, S3CannedACL acl, CancellationToken cancellationToken = default)
+        {
+            var putObjectRequest = new PutObjectRequest
+            {
+                Key = key, InputStream = stream, BucketName = _options.BucketName, CannedACL = acl
```
Comment (mingazhev) [inline HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs:40]: Please, put every property initialization on separate line

File: HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+                var objectToUpload = objects[nextTaskIndex];
+                processingTasks.Add(Add(objectToUpload.key, objectToUpload.stream, cancellationToken));
+            }
+
+            while (processingTasks.Count > 0)
+            {
+                var task = await Task.WhenAny(processingTasks);
+                processingTasks.Remove(task);
+
+                processedTasks.Add(await task);
```
Comment (mingazhev) [inline HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs:97]: Is await needed there?

File: HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs (hunk 3 above)
Comment (mingazhev) [reply HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs:239]: I don't think so, these operations are not processor-bound

File: HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs (hunk 8 above)
Comment (oleg1std) [reply HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs:97]: An inappropriate name for the list of results.

File: HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs (hunk 3 above)
Comment (oleg1std) [reply HappyTravel.AmazonS3Client/Services/AmazonS3ClientService.cs:239]: It's the number of concurrent uploads to a bucket

**Review by kirillta — APPROVED** [review]

**Pull request merged**